pkg/diagnostics/networkpod/pod.go

```diff
@@ -0,0 +1,288 @@
+package network
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	kcontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	kexec "k8s.io/kubernetes/pkg/util/exec"
+
+	osclient "github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/diagnostics/types"
+	"github.com/openshift/origin/pkg/sdn/api"
+	sdnplugin "github.com/openshift/origin/pkg/sdn/plugin"
+	"github.com/openshift/origin/pkg/util/netutils"
+)
+
+const (
+	CheckPodNetworkName = "CheckPodNetwork"
+)
+
+// CheckPodNetwork is a Diagnostic to check communication between pods in the cluster.
+type CheckPodNetwork struct {
+	KubeClient *kclient.Client
+	OSClient   *osclient.Client
+}
+
+// Name is part of the Diagnostic interface and just returns name.
+func (d CheckPodNetwork) Name() string {
+	return CheckPodNetworkName
+}
+
+// Description is part of the Diagnostic interface and just returns the diagnostic description.
+func (d CheckPodNetwork) Description() string {
+	return "Check pod to pod communication in the cluster. In case of ovs-subnet network plugin, all pods should be able to communicate with each other and in case of multitenant network plugin, pods in non-global projects should be isolated and pods in global projects should be able to access any pod in the cluster and vice versa."
+}
+
+// CanRun is part of the Diagnostic interface; it determines if the conditions are right to run this diagnostic.
+func (d CheckPodNetwork) CanRun() (bool, error) {
+	if d.KubeClient == nil {
+		return false, errors.New("must have kube client")
+	} else if d.OSClient == nil {
+		return false, errors.New("must have openshift client")
+	}
+	return true, nil
+}
+
+// Check is part of the Diagnostic interface; it runs the actual diagnostic logic
+func (d CheckPodNetwork) Check() types.DiagnosticResult {
+	r := types.NewDiagnosticResult(CheckPodNetworkName)
+
+	pluginName, ok, err := getOpenShiftNetworkPlugin(d.OSClient)
+	if err != nil {
+		r.Error("DPodNet1001", err, fmt.Sprintf("Checking network plugin failed. Error: %s", err))
+		return r
+	}
+	if !ok {
+		r.Warn("DPodNet1002", nil, fmt.Sprintf("Skipping pod connectivity test. Reason: Not using openshift network plugin."))
+		return r
+	}
+
+	localPods, nonlocalPods, err := getLocalAndNonLocalPods(d.KubeClient)
+	if err != nil {
+		r.Error("DPodNet1003", err, fmt.Sprintf("Getting local and nonlocal pods failed. Error: %s", err))
+		return r
+	}
+
+	vnidMap := map[string]uint32{}
```

Comment:
it would make later checks clearer if you had vnidMap be nil for single-tenant, and non-nil for multi-tenant, rather than checking the length